From: Thomas Preud'homme <thomas.preudhomme@celest.fr>
Date: Sun, 11 Apr 2010 01:53:40 +0200
Subject: [PATCH] 419203_fix_sizeof_parse_error.diff

---
 tccgen.c |   25 +++++++++++++++++++------
 1 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/tccgen.c b/tccgen.c
index b4b5d59..e04297b 100644
--- a/tccgen.c
+++ b/tccgen.c
@@ -47,6 +47,13 @@ void vsetc(CType *type, int r, CValue *vc)
     vtop->c = *vc;
 }
 
+/* push constant of type "type" with useless value */
+void vpush(CType *type)
+{
+    CValue cval;
+    vsetc(type, VT_CONST, &cval);
+}
+
 /* push integer constant */
 void vpushi(int v)
 {
@@ -2944,11 +2951,14 @@ static void vpush_tokc(int t)
 
 static void unary(void)
 {
-    int n, t, align, size, r;
+    int n, t, align, size, r, sizeof_caller;
     CType type;
     Sym *s;
     AttributeDef ad;
+    static int in_sizeof = 0;
 
+    sizeof_caller = in_sizeof;
+    in_sizeof = 0;
     /* XXX: GCC 2.95.3 does not generate a table although it should be
        better here */
  tok_next:
@@ -3045,6 +3055,10 @@ static void unary(void)
                 memset(&ad, 0, sizeof(AttributeDef));
                 decl_initializer_alloc(&type, &ad, r, 1, 0, 0);
             } else {
+                if (sizeof_caller) {
+                    vpush(&type);
+                    return;
+                }
                 unary();
                 gen_cast(&type);
             }
@@ -3114,11 +3128,8 @@ static void unary(void)
     case TOK_ALIGNOF2:
         t = tok;
         next();
-        if (tok == '(') {
-            parse_expr_type(&type);
-        } else {
-            unary_type(&type);
-        }
+        in_sizeof++;
+        unary_type(&type); // Perform a in_sizeof = 0;
         size = type_size(&type, &align);
         if (t == TOK_SIZEOF) {
             if (size < 0)
@@ -3735,9 +3746,11 @@ static void expr_type(CType *type)
 static void unary_type(CType *type)
 {
     int a;
+    void *vtop_saved;
 
     a = nocode_wanted;
     nocode_wanted = 1;
+    vtop_saved = vtop;
     unary();
     *type = vtop->type;
     vpop();
-- 
