--- tcc.orig/configure	2007-08-15 19:15:52.000000000 +0000
+++ tcc/configure	2007-08-15 19:19:22.000000000 +0000
@@ -31,7 +31,6 @@
 host_cc="gcc"
 ar="ar"
 make="make"
-strip="strip"
 cpu=`uname -m`
 case "$cpu" in
   i386|i486|i586|i686|i86pc|BePC|i686-AT386)
@@ -134,7 +133,6 @@
 
 cc="${cross_prefix}${cc}"
 ar="${cross_prefix}${ar}"
-strip="${cross_prefix}${strip}"
 
 if test "$mingw32" = "yes" ; then
     LIBSUF=".lib"
@@ -294,7 +292,6 @@
 echo "#define GCC_MAJOR $gcc_major" >> $TMPH
 echo "HOST_CC=$host_cc" >> config.mak
 echo "AR=$ar" >> config.mak
-echo "STRIP=$strip -s -R .comment -R .note" >> config.mak
 echo "CFLAGS=$CFLAGS" >> config.mak
 echo "LDFLAGS=$LDFLAGS" >> config.mak
 echo "LIBSUF=$LIBSUF" >> config.mak
--- tcc.orig/Makefile	2007-08-15 20:24:57.000000000 +0000
+++ tcc/Makefile	2007-08-15 20:25:16.000000000 +0000
@@ -204,7 +204,7 @@
 
 tcc_install: $(PROGS) tcc.1 libtcc1.a $(BCHECK_O) tcc-doc.html
 	mkdir -p "$(bindir)"
-	$(INSTALL) -s -m755 $(PROGS) "$(bindir)"
+	$(INSTALL) -m755 $(PROGS) "$(bindir)"
 ifndef CONFIG_WIN32
 	mkdir -p "$(mandir)/man1"
 	$(INSTALL) tcc.1 "$(mandir)/man1"
