From: Romain Francoise <rfrancoise@debian.org>
Date: Sat, 13 Mar 2010 18:07:11 +0100
Subject: [PATCH] buildsys.diff

---
 Makefile |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index 730b817..17d3e98 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 TOP ?= .
 include $(TOP)/config.mak
 
-CFLAGS+=-g -Wall
+CFLAGS+=-O2 -g -Wall
 CFLAGS_P=$(CFLAGS) -pg -static -DCONFIG_TCC_STATIC
 LIBS_P=
 
@@ -184,7 +184,7 @@ endif
 endif
 
 %.o: %.c
-	$(LIBTCC1_CC) -o $@ -c $< -O2 -Wall
+	$(LIBTCC1_CC) -o $@ -c $< $(CFLAGS)
 
 %.o: %.S
 	$(LIBTCC1_CC) -o $@ -c $<
@@ -193,7 +193,7 @@ libtcc1.a: $(LIBTCC1_OBJS)
 	$(AR) rcs $@ $^
 
 bcheck.o: bcheck.c
-	$(CC) -o $@ -c $< -O2 -Wall
+	$(CC) -o $@ -c $< $(CFLAGS)
 
 # install
 TCC_INCLUDES = stdarg.h stddef.h stdbool.h float.h varargs.h tcclib.h
-- 
